t/js_saml.t

```diff
@@ -0,0 +1,1642 @@
+            digest_saml($signature_element[0], true);
+            signature_saml($signature_element[0], $idp_priv, true);
+        } else {
+            $signature_element[0]->parentNode->
+                removeChild($signature_element[0]);
+        }
+    } elsif ($type eq 'LogoutRequest') {
+        $msg->setAttribute('Destination', $cfg->{saml_sp_slo_url});
+
+        # Subject processing
+        my ($nameid_element) = $xpc->findnodes('//saml:NameID');
+        $nameid_element->setAttribute('SPNameQualifier',
+            $cfg->{saml_idp_entity_id});
+
+        # Signature processing
+        if ($cfg->{saml_sp_want_signed_slo} eq 'true') {
+            digest_saml($signature_element[0], true);
+            signature_saml($signature_element[0], $idp_priv, true);
+        } else {
+            $signature_element[0]->parentNode->
+                removeChild($signature_element[0]);
+        }
+    } else {
+        die "Unknown SAML message type: $type";
+    }
+
+    return $xml_obj;
+}
+
+sub send_saml {
+    my ($xml_obj, $dst, %extra) = @_;
+    my ($r, $b64);
+
+    $dst //= $acs;
+
+    $XML::LibXML::skipXMLDeclaration = 1;
+    my $xml_str = $xml_obj->toString();
+
+    if (exists($extra{method}) && $extra{method} eq 'get') {
+        my $compressed_xml;
+        rawdeflate(\$xml_str => \$compressed_xml)
+            or die "rawdeflate failed: $RawDeflateError\n";
+        $b64 = encode_base64($compressed_xml, '');
+        my $url = $dst . "?SAMLResponse=" . uri_escape($b64) .
+            ($extra{relay_state} ? "&RelayState=" . $extra{relay_state} : "");
+        $r = get($url, %extra);
+    } else {
+        $b64 = encode_base64($xml_str, '');
+        my $body = "SAMLResponse=" . uri_escape($b64) .
+            ($extra{relay_state} ? "&RelayState=" . $extra{relay_state} : "");
+        $r = http_post($dst, $body, %extra);
+    }
+
+    return $r;
+}
+
+sub cfg_post {
+    my ($arg, $post, $host) = @_;
+    $host //= 'sp.example.com';
+    my $json;
+
+    if (ref $arg eq 'HASH') {
+        $json = $arg;
+    } elsif (!ref $arg) {
+        my ($key, $value) = each %{decode_json($arg)};
+        $cfg->{$key} = $value;
+        $json = {$key => $value};
+    } else {
+        die "Invalid arguments for cfg_post";
+    }
+
+    for my $key (keys %$json) {
+        my $value = $json->{$key};
+        my $data = { $host => $value };
+        my $data_string = encode_json($data);
+
+        if ($post) {
+            http_post("$kv/$key", $data_string);
+        } else {
+            http_patch("$kv/$key", $data_string);
+        }
+    }
+}
+
+sub cfg_verify {
+    my ($param, $test_name) = @_;
+
+    my $url = "$kv/$param";
+    my $original_value = getkv($url);
+
+    http_patch($url, '{"sp.example.com": ""}');
+    my $r = get('/login');
+    http_patch($url, $original_value);
+
+    like($r, qr/(?=.*Invalid)(?=.*$param)/s, $test_name);
+
+    return $r;
+}
+
+sub http_post {
+	my ($url, $body, %extra) = @_;
```

Comment:
Fix tabs

Reply:
Oh, thanks for the note, actually I was horrified by how many incorrect indents were in the code...